clients/src/main/java/org/apache/kafka/clients/consumer/internals/MetadataErrorManager.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.Metadata;
+import org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.PollResult;
+import org.apache.kafka.clients.consumer.internals.events.BackgroundEventHandler;
+import org.apache.kafka.clients.consumer.internals.events.ErrorEvent;
+
+public class MetadataErrorManager implements RequestManager {
+    private final Metadata metadata;
+    private final BackgroundEventHandler backgroundEventHandler;
+
+    public MetadataErrorManager(Metadata metadata, BackgroundEventHandler backgroundEventHandler) {
+        assert metadata != null;
+        assert backgroundEventHandler != null;
+
+        this.metadata = metadata;
+        this.backgroundEventHandler = backgroundEventHandler;
+    }
+
+    @Override
+    public PollResult poll(long currentTimeMs) {
+        maybePropagateMetadataError();
+
+        return PollResult.EMPTY;
+    }
+
+    private void maybePropagateMetadataError() {
+        try {
+            metadata.maybeThrowAnyException();
+        } catch (Exception e) {
+            backgroundEventHandler.add(new ErrorEvent(e));
+        }
```

Comment:
Ideally these five lines could just be in `ConsumerNetworkThread`, either inline in `runOnce()` or as a separate method.

Reply:
I agree.
If these five lines be in `ConsumerNetworkThread`, I think the drawbacks of the simple version mentioned by @philipnee can also be addressed.

Reply:
@philipnee How about these five lines be in `ConsumerNetworkThread`?
I will add integration test

Reply:
hey @appchemist - seems like you closed this pr. do you mean to continue working on this PR?

Reply:
hey, @philipnee - No, If you agree, I will close this PR

Reply:
thanks, is this the PR to review? https://github.com/apache/kafka/pull/16043

Reply:
@philipnee right, please take a look.(https://github.com/apache/kafka/pull/16043)
Can i close this pr(https://github.com/apache/kafka/pull/15961)?